metals/src/main/scala/scala/meta/internal/metals/FoldingRangeProvider.scala

```diff
@@ -0,0 +1,65 @@
+package scala.meta.internal.metals
+
+import java.util
+
+import org.eclipse.lsp4j.{FoldingRange, FoldingRangeCapabilities}
+
+import scala.meta.Term.Block
+import scala.meta.internal.metals.FoldingMode._
+import scala.meta.internal.metals.FoldingRanges.{Factory, Region}
+import scala.meta.io.AbsolutePath
+import scala.meta.transversers.SimpleTraverser
+import scala.meta.{Position, Tree}
+
+final class FoldingRangeProvider(val trees: Trees, mode: FoldingMode) {
+  def getRangedFor(path: AbsolutePath): util.List[FoldingRange] = {
+    trees
+      .get(path)
+      .map(findFoldingRanges)
+      .getOrElse(util.Collections.emptyList())
+  }
+
+  private def findFoldingRanges(tree: Tree): util.List[FoldingRange] = {
+    val ranges = new util.ArrayList[FoldingRange]()
+    new RangeFinder(ranges).apply(tree)
+    ranges
+  }
+
+  private class RangeFinder(ranges: util.ArrayList[FoldingRange])
```

Comment:
Can this be replaced with?

```scala
val ranges = tree.collect {
  case block @ Term.Block(_) => createRange(block.pos)
}
```

See https://scalameta.org/docs/trees/guide.html#simple-traversals for more examples how to use `collect`. `Traverser` should only be needed for more custom traversals, for example when we don't want to visit all tree nodes.